### DISCUSSION TITLE
Use Moq to make mocking easier in Unit-Tests

### DISCUSSION THREAD

**Discussion by ramonB1996** [CommunityToolkit/Maui#1117] [New Feature Discussions]
Hi,

I would love to know if `Moq` (https://github.com/moq/moq4) mocking library has ever been considered to be added to the `Maui.CommunityToolkit`? 

If not, this would be the discussion section for this.

The problem: At the moment we have to create extra classes just to use them as a Mock, as far as my understanding goes. 

All opinions welcome of course, I am willing to do some PR's for Unit tests and want to make it easier for myself and others if possible.

**Comment by bijington** [comment]
My opinion of Moq varies quite a bit. I have used it extensively in other projects and it has been great but later on I find it's usage results in more difficult to follow code/tests.

While on the topic of mocking I would like to advocate for making our current mocks more descriptive so rather than writing something like:

```csharp
var mockConverter = new MockValueConverter(value => true);
```

We could modify it to something like:

```csharp
var mockConverter = MockValueConverter.ThatReturns(true, for: value);
```

You don't easily get this readability from Moq but you could wrap it underneath.

Also I believe a number of our mocks have been lifted from the .NET MAUI repo.

That isn't me saying I'm against it but I don't see a big enough reason to do it. I am happy to be persuaded otherwise though 🙂

**Comment by VladislavAntonyuk** [comment]
I prefer NSubstitute. It has a much cleaner syntax. However, we decided to use Fake implementation to test without 3rd party libraries

**Comment by ramonB1996** [comment]
I understand your decision. I have 2 question then: 

To test a class like `UseCommunityToolkitAnalyzer`, what would you do for these parts:

1. Input that the method `Initialize()` needs. How would you mock that away?
2. Check if certain methods were called or were not called? In `Moq` you can use `verify()` to check this.

I can live with the fact that we do not add this library, as it would increase package size as well. But I would love to know how to contribute to the unit-tests in these instances in a proper way.

**Reply by pictos** [reply]
@ramonB1996 testing Roslyn analyzer needs other methods, you can create a compilation with a string code and test if that works. You can find more examples on the Roslyn [repo](https://github.com/dotnet/roslyn/blob/main/src/Analyzers/CSharp/Tests/AddBraces/AddBracesFixAllTests.cs), since we need a workload in order to make it work that could be more tricky, to implement a unit test... If that looks hard, I would suggest to implement unit tests for our Source Generator, it will follow almost the same idea as the Analyzer unit test

**Comment by CliffAgius** [comment]
Closed as answered
